### ISSUE TITLE
Duck typing errors in generic code should include "called from here" trace

### ISSUE THREAD

**Issue by ghost** [ziglang/zig#1270]
See the comments in this code paste:
```
fn container(comptime T: type) type {
  return struct {
    const Self = this;

    value: T,

    // note that this method uses duck typing (assumes T contains a u32)
    fn getContents(self: *Self) u32 {
      return self.value.contents; // this is where the compile error will happen
    }
  };
}

const Something = struct{
  contents: u32,
};

const SomethingElse = struct{
  contents: f32,
};

test "" {
  var a: container(Something) = undefined;
  // struct contains a u32 so this will work.
  _ = a.getContents();

  var b: container(SomethingElse) = undefined;
  // struct contains a f32 so this will not work. this is where the actual mistake is
  _ = b.getContents();
}
```

This fails with a compile error that only mentions the single line where the compiler ran into a wall.

```
/home/dbandstra/oxid/test.zig:9:24: error: expected type 'u32', found 'f32'
      return self.value.contents;
```

It's not clear from looking at line 9 in the code what the compiler is talking about. It mentions 'f32' but there's no f32 to be seen there.

The compiler should display a call trace ("called from here") to help the human realize where the bad call was made.

**Comment by dbandstra** [comment]
A smaller test case of what I assume is the same problem:

```
fn func(p: var) @typeInfo(@typeOf(p)).Pointer.child {
  unreachable;
}

test "" {
  _ = func(1);
}
```

Output:

```
/home/dbandstra/oxid/test.zig:1:38: error: accessing union field 'Pointer' while field 'ComptimeInt' is set
fn func(p: var) @typeInfo(@typeOf(p)).Pointer.child {
                                     ^
```
On the other hand, at least in this example, if the error happens inside of the function body instead of in the return type, the compiler provides a fuller trace. 
```
fn func(p: var) i32 {
  _ = @typeInfo(@typeOf(p)).Pointer.child;
  unreachable;
}

test "" {
  _ = func(1);
}
```
Output:
```
/home/dbandstra/oxid/test.zig:2:28: error: accessing union field 'Pointer' while field 'ComptimeInt' is set
  _ = @typeInfo(@typeOf(p)).Pointer.child;
                           ^
/home/dbandstra/oxid/test.zig:7:11: note: called from here
  _ = func(1);
          ^
```

**Comment by dbandstra** [comment]
An even smaller test case I ran into today, using `std.math.min`:

```zig
const std = @import("std");

test "" {
  const a: usize = 1;
  const b: i32 = 2;
  _ = std.math.min(a, b);
}
```

Output trace doesn't include anything from my actual source file:

```
/home/dbandstra/zig/build/lib/zig/std/math.zig:245:38: error: incompatible types: 'usize' and 'i32'
pub fn min(x: var, y: var) @typeOf(x + y) {
                                     ^
/home/dbandstra/zig/build/lib/zig/std/math.zig:245:28: note: called from here
pub fn min(x: var, y: var) @typeOf(x + y) {
                           ^
/home/dbandstra/zig/build/lib/zig/std/math.zig:245:36: note: type 'usize' here
pub fn min(x: var, y: var) @typeOf(x + y) {
                                   ^
/home/dbandstra/zig/build/lib/zig/std/math.zig:245:40: note: type 'i32' here
pub fn min(x: var, y: var) @typeOf(x + y) {
                                       ^
```

**Comment by andrewrk** [comment]
solved by reference traces in #12796
